### PR TITLE
Hclang 62 explain generics and overloading

### PR DESCRIPTION
**Pull request by technight**
### Description
Create a section about overloading and generics within HCL.

### Changelog
 - Added overloading section
 - Added generics section

### Issues solved 
 - HCLANG-62

### Issues opened as a result of this 
Add something here if new issues arise because of this. For instance if new features-requests or bugs arise from this PR. If this is not the case, then remove this section.

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: report/3.Theory/LanguageSpecification.tex [hunk 1]
```diff
@@ -93,4 +93,72 @@ \section{Language Specification}
...
+The language will include function overloading. 
+This means that multiple functions can share the same name and have different implementations based on the different parameters that the function accepts. 
+An example could be the plus function. 
+The plus function could accept two numbers or two strings, but the implementation of the plus function would depend on the type.
+If two numbers were passed, the obvious result should be the sum of the two numbers.
+If on the other hand two strings were passed, the function could be implemented in such a way that the result would be the concatenation of the two strings.
+
+The function overloading in HCL is however a bit limited in that functions can not be overloaded with different amounts of parameters, only different parameter types. 
+This is due to the nature of the syntax in HCL where function parameters are not directly coupled to a function. 
+Therefore it would be impossible for the compiler to figure out what is parameters and what is functions.
```
Comment (Abildgaard96) [inline report/3.Theory/LanguageSpecification.tex]: Determine instead of 'figure out'

File: report/3.Theory/LanguageSpecification.tex [hunk 2]
```diff
@@ -93,4 +93,72 @@ \section{Language Specification}
...
+Consider this example:
+
+\begin{lstlisting}
+1 to 10 sum print
+\end{lstlisting}
+
+Obviously the sum function accepts a list that should be summed. 
+However, if the sum function could potentially also accept a modifier function, that were to alter the result of the sum function, it would be impossible for compiler to figure out whether print was that modifier function, or the sum should be passed as a parameter to the print function.
```
Comment (Abildgaard96) [inline report/3.Theory/LanguageSpecification.tex]: not exactly true, as we specifically use :: infront of functions that are used as parameters.
also add a 'the' before 'compiler' at "impossible for [the] compiler..."

File: report/3.Theory/LanguageSpecification.tex [hunk 3]
```diff
@@ -93,4 +93,72 @@ \section{Language Specification}
...
+
+Function overloading requires nothing special of the user, just declare two functions with the same amount of parameters, but with different types.
```
Comment (Abildgaard96) [inline report/3.Theory/LanguageSpecification.tex]: evt add "as seen in listing below"

File: report/3.Theory/LanguageSpecification.tex [hunk 4]
```diff
@@ -93,4 +93,72 @@ \section{Language Specification}
...
+\begin{lstlisting}
+var isAwesome = (string name) -> bool {
+	return name equals "HCL" thenElse {true} {false}
+}
+var isAwesome = (num number) -> bool {
+	return number equals 42 thenElse {true} {false}
+}
+
+42 isAwesome // True
+"Java" isAwesome // False
```
Comment (Abildgaard96) [inline report/3.Theory/LanguageSpecification.tex]: shots fired!
 - not actual comment, just wanted to give credit where credit is due..

File: report/3.Theory/LanguageSpecification.tex [hunk 5]
```diff
@@ -93,4 +93,72 @@ \section{Language Specification}
...
+	elementsToMap onEach {mappedValues add (value mapFunction)}
+	return mappedValues
+}
+\end{lstlisting}
+
+Generic types in the above example is T1 and T2. 
+Since nothing type specific ever happens with these types, we can handle all types the same, whether that be a string, a number or some other type will not matter.
+
+
+% Should the above textbf be sub sub sections?
```
Comment (Abildgaard96) [inline report/3.Theory/LanguageSpecification.tex]: kan personligt bedst li' dem som bare textbf

File: report/3.Theory/LanguageSpecification.tex [hunk 6]
```diff
@@ -93,4 +93,69 @@ \section{Language Specification}
...
+	return number equals 42 thenElse {true} {false}
+}
+
+42 isAwesome // True
+"Java" isAwesome // False
+\end{lstlisting}
+
+The main reasoning for including function overloading is simplicity for the user.
+For instance for printing. 
+The print function should be able to print all basic types, so that the user does not have to worry about what type he is about to print.
```
Comment (nillas12) [inline report/3.Theory/LanguageSpecification.tex]: Change "about what type he is about to print." to worry about the type of what he is about to print."

**Pull request merged**